python-package/lightgbm/sklearn.py

```diff
@@ -874,7 +874,7 @@ def n_estimators_(self) -> int:
         This might be less than parameter ``n_estimators`` if early stopping was enabled or
         if boosting stopped early due to limits on complexity like ``min_gain_to_split``.
         """
-        if not self.__sklearn_is_fitted__():
+        if not self.__sklearn_is_fitted__() or self._Booster is None:
```

Comment:
`__sklearn_is_fitted__()` was introduced with the aim to avoid a such type of checks for initial values of attributes: #4654. I don't think that workaround for mypy error worth introducing nonsense code. I'm not supporting this particular change.

Reply:
Consider the following:

```python
import lightgbm as lgb
from sklearn.datasets import load_breast_cancer

X, y = load_breast_cancer(return_X_y=True)
data = lgb.Dataset(X, label=y)

# train model
clf = lgb.LGBMClassifier(
    objective="binary",
    n_estimators=5
)

clf.fit(X, y)
clf._Booster = None
clf.n_estimators_
```

On `master`, that code will yield this error:

> AttributeError: 'NoneType' object has no attribute 'current_iteration'

As of this PR, it would return this error

> No n_estimators found. Need to call fit beforehand.

So this is not a "false positive" from `mypy`. `mypy` is correctly complaining about the fact that this code path and others like it treat `self._Booster` as non-None even though it is possible for it to be `None`.

This is also why I don't see the proposed change here as "nonsense". With the change proposed on this line, in the unlikely and unsupported but **possible** scenario where `_Booster` has been set to `None`, you'll get a more informative error message.

--------------

Anyway, I'll remove this and use `# type: ignore` comments. But wanted you to understand that I wasn't just trying to do weird things to silence a linter.

Reply:
updated in https://github.com/microsoft/LightGBM/pull/4837/commits/6e3420187f1115a7690a03b81fc417cbe28c6e76

Reply:
> `clf._Booster = None`

Accessing private attributes, especially setting them to a custom value, from an outside code shouldn't happen in a normal scenario. I think it sounds quite weird to set private attribute to a random value and expect correct behavior. It's likely the same as
```python
...

clf._Booster = 42
# expect something meaningful
clf.predict(X)
```

As we set all private attributes in a `fit()` method and users shouldn't modify them, I still believe we shouldn't add more checks and use only `__sklearn_is_fitted__()`.

Reply:
for sure, this is why I thought of this situation as "unlikely and unsupported but possible". I'm ok with not adding these checks and just using `# type: ignore`.